File: tools/improve/tests/original/maudec/py/free-simple.py
import enum

class Foo(enum.Enum):
	Int = 0  # int
	Float = 1  # float
	Err = 2  # 


def pi() -> float:
	return 3.141592653589793


def _u003du0060u005b_u0060u005d_i(X: float, Z: float, Y: float) -> bool:
	return abs((X - Y)) < Z


def square(a1: Foo) -> Foo:
	match a1:
		case (Foo.Int, N):
			return (Foo.Int, N * N)

		case (Foo.Float, F):
			return (Foo.Float, F * F)



def isFloat(O: Foo) -> bool:
	match O:
		case (Foo.Float, _):
			return True

		case _:
			return False



def getFloat(a1: Foo) -> float:
	match a1:
		case (Foo.Float, F):
			return F



def nat2float(a1: int) -> float:
	match a1:
		case 0:
			return 0.0

		case K if K >= 1:
			return 1.0 + nat2float(K - 1)



def add(a1: Foo, a2: Foo) -> Foo:
	match (a1, a2):
		case ((Foo.Int, N1), (Foo.Int, N2)):
			return (Foo.Int, N1 + N2)

		case ((Foo.Float, F1), (Foo.Float, F2)):
			return (Foo.Float, F1 + F2)

		case ((Foo.Int, N1), (Foo.Float, F2)):
			return (Foo.Float, nat2float(N1) + F2)

		case ((Foo.Float, F1), (Foo.Int, N2)):
			return (Foo.Float, F1 + nat2float(N2))

File: tools/improve/tests/original/maudec/cpp/free-simple.cpp
#include <cmath>
#include <variant>


struct Foo : public std::variant<
	int,
	float,
	std::monostate
> {
	using std::variant<int, float, std::monostate>::variant;

	enum Variant {
		INT,
		FLOAT,
		ERR,
	};
};

float pi() {
	return 3.141592653589793;
}

bool _u003du0060u005b_u0060u005d_(float X, float Z, float Y) {
	return abs(X - Y) < Z;
}

Foo square(const Foo& a1) {
	if (a1.index() == Foo::INT) {
		return Foo(std::in_place_index<Foo::INT>, std::get<Foo::INT>(a1) * std::get<Foo::INT>(a1));
	}
	if (a1.index() == Foo::FLOAT) {
		return Foo(std::in_place_index<Foo::FLOAT>, std::get<Foo::FLOAT>(a1) * std::get<Foo::FLOAT>(a1));
	}
}

bool isFloat(const Foo& O) {
	if (O.index() == Foo::FLOAT) {
		return true;
	}
	return false;
}

float getFloat(const Foo& a1) {
	if (a1.index() == Foo::FLOAT) {
		return std::get<Foo::FLOAT>(a1);
	}
}

float nat2float(int a1) {
	if (a1 == 0) {
		return 0.0;
	}
	if (a1 >= 1) {
		return 1.0 + nat2float(a1 - 1);
	}
}

Foo add(const Foo& a1, const Foo& a2) {
	if (a1.index() == Foo::INT && a2.index() == Foo::INT) {
		return Foo(std::in_place_index<Foo::INT>, std::get<Foo::INT>(a1) + std::get<Foo::INT>(a2));
	}
	if (a1.index() == Foo::FLOAT && a2.index() == Foo::FLOAT) {
		return Foo(std::in_place_index<Foo::FLOAT>, std::get<Foo::FLOAT>(a1) + std::get<Foo::FLOAT>(a2));
	}
	if (a1.index() == Foo::INT && a2.index() == Foo::FLOAT) {
		return Foo(std::in_place_index<Foo::FLOAT>, nat2float(std::get<Foo::INT>(a1)) + std::get<Foo::FLOAT>(a2));
	}
	if (a1.index() == Foo::FLOAT && a2.index() == Foo::INT) {
		return Foo(std::in_place_index<Foo::FLOAT>, std::get<Foo::FLOAT>(a1) + nat2float(std::get<Foo::INT>(a2)));
	}
}

